backend/services/implementations/classService.ts

```diff
@@ -0,0 +1,260 @@
+      Logger.error(
+        `Failed to update class. Reason = ${getErrorMessage(error)}`,
+      );
+      throw error;
+    }
+    return (await this.mapClassToClassDTOs([updatedClass]))[0];
+  }
+
+  async deleteClass(id: string): Promise<string> {
+    try {
+      const deletedClass: Class | null = await MgClass.findByIdAndDelete(id);
+      if (!deletedClass) {
+        throw new Error(`Class with id ${id} not found`);
+      }
+      return id;
+    } catch (error: unknown) {
+      Logger.error(
+        `Failed to delete class. Reason = ${getErrorMessage(error)}`,
+      );
+      throw error;
+    }
+  }
+
+  async createStudent(
+    student: StudentRequestDTO,
+    classId: string,
+  ): Promise<ClassResponseDTO> {
+    let newStudent: StudentResponseDTO;
+    let classObj: Class | null;
+
+    try {
+      // create a new StudentResponseDTO
+      newStudent = {
+        id: "",
+        firstName: student.firstName,
+        lastName: student.lastName,
+        studentNumber: student.studentNumber,
+      };
+
+      classObj = await MgClass.findById(classId);
+
+      if (!classObj) {
+        throw new Error(`Class id ${classId} not found`);
+      }
+
+      classObj?.students.push(newStudent);
+      await classObj?.save();
+    } catch (error: unknown) {
+      Logger.error(
+        `Failed to create student. Reason = ${getErrorMessage(error)}`,
+      );
+      throw error;
+    }
+    return (await this.mapClassToClassDTOs([classObj]))[0];
+  }
+
+  async updateStudent(
+    studentId: string,
+    classId: string,
+    student: StudentRequestDTO,
+  ): Promise<ClassResponseDTO> {
+    let studentToUpdate;
+    let updatedClass: Class | null;
+    try {
+      updatedClass = await MgClass.findById(classId);
+
+      if (!updatedClass) {
+        throw new Error(`Class ${classId} not found`);
+      }
+
+      studentToUpdate = updatedClass.students.find(
+        (studentObj) => studentObj.id === studentId,
+      );
+
+      if (!studentToUpdate) {
+        throw new Error(
+          `Student id ${studentId} not found in class ${classId}`,
+        );
+      }
+
+      Object.assign(studentToUpdate, student);
+
+      await updatedClass.save();
+    } catch (error: unknown) {
+      Logger.error(
+        `Failed to update student. Reason = ${getErrorMessage(error)}`,
+      );
+      throw error;
+    }
+    return (await this.mapClassToClassDTOs([updatedClass]))[0];
+  }
+
+  async deleteStudent(studentId: string, classId: string): Promise<string> {
+    try {
+      const result = await MgClass.findByIdAndUpdate(
+        classId,
+        {
+          $pull: { students: { _id: studentId } },
+        },
+        { new: true, runValidators: true },
+      );
+
+      if (!result) {
+        throw new Error(`Class with id ${classId} not found`);
+      }
+
+      return studentId;
+    } catch (error: unknown) {
+      Logger.error(
+        `Failed to delete student from class. Reason = ${getErrorMessage(
```

Comment:
can we add the student and class ids into all the error messages in the service?